src/native/libs/System.Globalization.Native/pal_collation.m

```diff
@@ -295,4 +295,49 @@ int32_t GlobalizationNative_EndsWithNative(const uint16_t* localeName, int32_t l
     }
 }
 
+int32_t GlobalizationNative_GetSortKeyNative(const uint16_t* localeName, int32_t lNameLength, const UChar* lpStr, int32_t cwStrLength,
+                                             uint8_t* sortKey, int32_t cbSortKeyLength, int32_t options)
+{
+    @autoreleasepool {
+        if (cwStrLength == 0)
+        {
+            if (sortKey == NULL)
+                sortKey = malloc(1);
+            sortKey[0] = '\0';
+            return 1;
+        }
+        NSString *sourceString = [NSString stringWithCharacters: lpStr length: cwStrLength];
+        NSString *sourceStringCleaned = RemoveWeightlessCharacters(sourceString);
+
+        NSLocale *locale = GetCurrentLocale(localeName, lNameLength);
+        NSStringCompareOptions comparisonOptions = options == 0 ? 0 : ConvertFromCompareOptionsToNSStringCompareOptions(options);
+
+        // Generate a sort key for the original string based on the locale
+        NSString *transformedString = [sourceStringCleaned stringByFoldingWithOptions:comparisonOptions locale:locale];
+
+        // Convert the string to UTF-8 representation
+        const char *utf8Bytes = [transformedString UTF8String];
+        NSData *dataToUse = nil;
+        NSUInteger utf8Length = 0;
+        if (utf8Bytes != NULL) {
+            utf8Length = [transformedString lengthOfBytesUsingEncoding:NSUTF8StringEncoding];
+            dataToUse = [NSData dataWithBytes:utf8Bytes length:utf8Length];
+        } else {
+            // Convert the string to UTF-16 representation
+            dataToUse = [transformedString dataUsingEncoding:NSUTF16StringEncoding];
+            utf8Length = ([dataToUse length] / sizeof(uint16_t)) * 2;
```

Comment:
I'm not sure how the else can be happening and assuming UTF16 doesn't seem safe to me.
Also UTF8 is a variable length encoding so could be 1-4bytes, calculating the length based on `sizeof(uint16_t)) * 2` doesn't seem safe to me.

Reply:
`else` block was added for `System.Collections.Tests` failure. There are cases when `[transformedString UTF8String]` is null for passed string (like this line https://github.com/dotnet/runtime/blob/main/src/libraries/System.Collections/tests/Generic/Dictionary/HashCollisionScenarios/OutOfBoundsRegression.cs#L177)

Reply:
hmm that might just happen for strings with invalid surrogate pairs: https://stackoverflow.com/questions/47130742/whats-an-example-of-an-nsstring-for-which-utf8string-returns-null

I wonder if we need to convert to utf8 at all, maybe we could just return the `transformedString` as utf16 to managed?

Reply:
I will try returning `transformedString` as utf16 to managed to see how tests will pass.

Reply:
If managed is only expecting a byte array or span then I think we don't need to do any conversion to utf8/utf16 at all, we can use something like this:

```objc
NSUInteger transformedStringBytes = [transformedString lengthOfBytesUsingEncoding: NSUTF16StringEncoding];
sortKey = malloc(transformedStringBytes);
NSRange range = NSMakeRange(0, [transformedString length]);
BOOL result = [transformedString getBytes:sortKey maxLength:transformedStringBytes usedLength:NULL encoding:NSUTF16StringEncoding options:0 range:range remainingRange:NULL];

// check result etc
```

Reply:
Thanks, refactored the function.